src/osd/OSDMap.cc

```diff
@@ -4487,106 +4488,210 @@ int OSDMap::calc_pg_upmaps(
+            continue;
+          }
+          auto upit = it->second.begin();
+          for (; upit != it->second.end() && upit->first != osd; upit++)
+            ;
+          if (upit == it->second.end()) {
+            // no luck
+            it++;
+            continue;
+          }
+          ldout(cct, 10) << " existing pg_upmap_items " << it->second
+                         << " remapped " << pg << " out of underfull"
+                         << " osd." << osd << ","
+                         << " will try un-remap " << pg << " first"
                          << dendl;
-          ceph_assert(tmp.pg_upmap_items.count(i.first) == 0);
-          tmp.pg_upmap_items[i.first] = i.second;
-          pending_inc->new_pg_upmap_items[i.first] = i.second;
-          ++num_changed;
+          for (auto& i : it->second) {
+            temp_pgs_by_osd[i.second].erase(pg);
+            temp_pgs_by_osd[i.first].insert(pg);
+          }
+          to_unmap.insert(pg);
+          break;
         }
-      } else {
-        ldout(cct, 10) << " failed to find further changes to make" << dendl;
-        break;
+        if (to_upmap.size())
+          break;
       }
     }
+
+    if (!(to_unmap.size() || to_upmap.size())) {
+      ldout(cct, 10) << " failed to find changes to make" << dendl;
+      break;
+    }
+
+    // test change, apply if change is good
+    float new_stddev = 0;
+    map<int,float> temp_osd_deviation;
+    multimap<float,int> temp_deviation_osd;
+    for (auto& i : temp_pgs_by_osd) {
+      // make sure osd is still there (belongs to this crush-tree)
+      ceph_assert(osd_weight.count(i.first));
+      float target = osd_weight[i.first] * pgs_per_weight;
+      float deviation = (float)i.second.size() - target;
+      ldout(cct, 20) << " osd." << i.first
+                     << "\tpgs " << i.second.size()
+                     << "\ttarget " << target
+                     << "\tdeviation " << deviation
+                     << dendl;
+      temp_osd_deviation[i.first] = deviation;
+      temp_deviation_osd.insert(make_pair(deviation, i.first));
+      new_stddev += deviation * deviation;
+    }
+    ldout(cct, 10) << " stddev " << stddev << " -> " << new_stddev << dendl;
+    if (new_stddev >= stddev) {
+      ldout(cct, 10) << " stddev is not dropping, break" << dendl;
+      break;
+    }
+
+    // hooray!
+    stddev = new_stddev;
+    pgs_by_osd = temp_pgs_by_osd;
+    osd_deviation = temp_osd_deviation;
+    deviation_osd = temp_deviation_osd;
+    for (auto& i : to_unmap) {
+      ldout(cct, 10) << " unmap pg " << i << dendl;
+      ceph_assert(tmp.pg_upmap_items.count(i));
+      tmp.pg_upmap_items.erase(i);
+      pending_inc->old_pg_upmap_items.insert(i);
+      ++num_changed;
+    }
+    for (auto& i : to_upmap) {
+      ldout(cct, 10) << " upmap pg " << i.first
+                     << " pg_upmap_items " << i.second
+                     << dendl;
+      ceph_assert(tmp.pg_upmap_items.count(i.first) == 0);
```

Comment:
@dvanders Sorry for the inconvenience. I forgot to remove this assert, which is a false alarm because attempting to modify an existing pg_upmap_items is legal now.

Manually killing off this assert or cherry-picking https://github.com/xiexingguo/ceph/commit/12bd6208374bab1240fa81217d4a63656bdb11b5 should make this patch work.

Thanks for the quick feedback!